org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/symbols/LSPSymbolInFileDialog.java

```diff
@@ -41,40 +37,42 @@
 import org.eclipse.swt.widgets.Shell;
 import org.eclipse.ui.dialogs.FilteredTree;
 import org.eclipse.ui.dialogs.PatternFilter;
+import org.eclipse.ui.internal.navigator.NavigatorContentService;
+import org.eclipse.ui.internal.navigator.NavigatorDecoratingLabelProvider;
+import org.eclipse.ui.navigator.CommonViewerSorter;
 import org.eclipse.ui.texteditor.ITextEditor;
 
 public class LSPSymbolInFileDialog extends PopupDialog {
 
 	private final OutlineViewerInput outlineViewerInput;
+	private final ITextEditor textEditor;
 
 	public LSPSymbolInFileDialog(@NonNull Shell parentShell, @NonNull ITextEditor textEditor,
 			@NonNull IDocument document, @NonNull LanguageServer languageServer) {
 		super(parentShell, PopupDialog.INFOPOPUPRESIZE_SHELLSTYLE, true, true, true, false, false, null, null);
 		outlineViewerInput = new OutlineViewerInput(document, languageServer, textEditor);
+		this.textEditor = textEditor;
 		create();
 	}
 
 	@Override
 	protected Control createDialogArea(Composite parent) {
 		IFile documentFile = outlineViewerInput.documentFile;
 		getShell().setText(NLS.bind(Messages.symbolsInFile, documentFile == null ? null : documentFile.getName()));
+
 		FilteredTree filteredTree = new FilteredTree(parent, SWT.BORDER, new PatternFilter(), true, false);
 		TreeViewer viewer = filteredTree.getViewer();
 
-		viewer.setContentProvider(new LSSymbolsContentProvider() {
-			@Override
-			public void inputChanged(Viewer viewer, Object oldInput, Object newInput) {
-				super.inputChanged(viewer, oldInput, newInput);
-				// eagerly populating the tree to avoid the initial 1-2 second delay
-				// when automatically populated by the internal ReconcilerOutlineUpdater
-				refreshTreeContentFromLS();
-			}
-		});
-		if (InstanceScope.INSTANCE.getNode(LanguageServerPlugin.PLUGIN_ID)
-				.getBoolean(CNFOutlinePage.SORT_OUTLINE_PREFERENCE, false)) {
-			viewer.setComparator(OutlineSorter.INSTANCE);
-		}
-		viewer.setLabelProvider(new SymbolsLabelProvider());
+		final var contentService = new NavigatorContentService(CNFOutlinePage.ID, viewer);
+		filteredTree.addDisposeListener(ev -> contentService.dispose());
+		final var contentProvider = contentService.createCommonContentProvider();
+		viewer.setContentProvider(contentProvider);
+
+		var sorter = new CommonViewerSorter();
+		sorter.setContentService(contentService);
+		viewer.setComparator(sorter);
```

Comment:
What does happen if alphabetical order is disabled? Is the order of elements (as provided by the model/LS) respected? It's important that we respect LS ordering by default. It would be great is some test get added about that.

Reply:
Yes. The OutlineSorter I implemented respects the current sort setting. If sorting is disabled (default) the compare funtion returns `0` meaning the items are shown in the order they are received from the LS.